src/Infrastructure.EntityFramework/Repositories/CollectionRepository.cs

```diff
@@ -769,4 +775,90 @@ private static async Task ReplaceCollectionUsersAsync(DatabaseContext dbContext,
         dbContext.CollectionUsers.RemoveRange(toDelete);
         // SaveChangesAsync is expected to be called outside this method
     }
+
+    public async Task CreateDefaultCollectionsAsync(Guid organizationId, IEnumerable<Guid> affectedOrgUserIds, string defaultCollectionName)
+    {
+        if (!affectedOrgUserIds.Any())
+        {
+            return;
+        }
+
+        using (var scope = ServiceScopeFactory.CreateScope())
+        {
+            var dbContext = GetDatabaseContext(scope);
+
+            var orgUserIdWithDefaultCollection = await GetOrgUserIdsWithDefaultCollectionAsync(dbContext, organizationId);
+
+            var missingDefaultCollectionUserIds = affectedOrgUserIds.Where(orgUserId => !orgUserIdWithDefaultCollection.Contains(orgUserId)).ToList();
+
+            var (collectionUsers, collections) = BuildDefaultCollectionForUsers(organizationId, missingDefaultCollectionUserIds, defaultCollectionName);
+
+            if (!collectionUsers.Any() || !collections.Any())
+            {
+                return;
+            }
+
+            await dbContext.BulkCopyAsync(collections);
+            await dbContext.BulkCopyAsync(collectionUsers);
+
+            await dbContext.SaveChangesAsync();
+        }
+    }
+
+    private async Task<HashSet<Guid>> GetOrgUserIdsWithDefaultCollectionAsync(DatabaseContext dbContext, Guid organizationId)
+    {
+        var results = await dbContext.OrganizationUsers
+                 .Where(ou => ou.OrganizationId == organizationId)
+                 .Join(
+                     dbContext.CollectionUsers,
+                     ou => ou.Id,
+                     cu => cu.OrganizationUserId,
+                     (ou, cu) => new { ou, cu }
+                 )
+                 .Join(
+                     dbContext.Collections,
+                     temp => temp.cu.CollectionId,
+                     c => c.Id,
+                     (temp, c) => new { temp.ou, Collection = c }
+                 )
+                 .Where(x => x.Collection.Type == CollectionType.DefaultUserCollection)
+                 .Select(x => x.ou.Id)
+                 .ToListAsync();
+
+        return results.ToHashSet();
```

Comment:
`ToListAsync` isn't required if you're going to convert it into a HashSet anyway.

Reply:
Unfortunately, there isn’t an async version of `ToHashSet`, so I just went with `ToListAsync` followed by ToHashSet. That way, the database call remains non-blocking, and we still get O(1) lookups later when filtering.

A `ToDictionaryAsync` could potentially work here, but `ToHashSet` feels more semantically appropriate since a dictionary is key-value pairs, and we’re only dealing with `OrgUserId`.

Reply:
Good point - there is a `ToHashSetAsync` but not until EF 9 and we're still on EF 8.x. I think this is a good solution in the meantime!